presto-geospatial-toolkit/src/test/java/com/facebook/presto/geospatial/TestGeometrySerialization.java

```diff
@@ -0,0 +1,204 @@
+        testSerialization("GEOMETRYCOLLECTION (MULTIPOLYGON (((30 20, 45 40, 10 40, 30 20))), GEOMETRYCOLLECTION (MULTIPOLYGON (((30 20, 45 40, 10 40, 30 20)))))");
+    }
+
+    @Test
+    public void testDeserializeEnvelope()
+    {
+        assertEquals(deserializeEnvelope(geometryFromText("MULTIPOINT (20 20, 25 25)")), new Envelope(20, 20, 25, 25));
+        assertEquals(deserializeEnvelope(geometryFromText("MULTILINESTRING ((1 1, 5 1), (2 4, 4 4))")), new Envelope(1, 1, 5, 4));
+        assertEquals(deserializeEnvelope(geometryFromText("POLYGON ((0 0, 0 4, 4 0))")), new Envelope(0, 0, 4, 4));
+        assertEquals(deserializeEnvelope(geometryFromText("MULTIPOLYGON (((0 0 , 0 2, 2 2, 2 0)), ((2 2, 2 4, 4 4, 4 2)))")), new Envelope(0, 0, 4, 4));
+        assertEquals(deserializeEnvelope(geometryFromText("GEOMETRYCOLLECTION (POINT (3 7), LINESTRING (4 6, 7 10))")), new Envelope(3, 6, 7, 10));
+        assertEquals(deserializeEnvelope(geometryFromText("POLYGON EMPTY")), null);
+        assertEquals(deserializeEnvelope(geometryFromText("POINT (1 2)")), new Envelope(1, 2, 1, 2));
+        assertEquals(deserializeEnvelope(geometryFromText("POINT EMPTY")), null);
+        assertEquals(deserializeEnvelope(geometryFromText("GEOMETRYCOLLECTION (GEOMETRYCOLLECTION (POINT (2 7), LINESTRING (4 6, 7 10)), POINT (3 7), LINESTRING (4 6, 7 10))")), new Envelope(2, 6, 7, 10));
+    }
+
+    private static void testSerialization(String wkt)
+    {
+        testEsriSerialization(wkt);
+        testJtsSerialization(wkt);
+    }
+
+    private static void testEsriSerialization(String wkt)
+    {
+        OGCGeometry expected = OGCGeometry.fromText(wkt);
+        OGCGeometry actual = deserialize(serialize(expected));
+        assertGeometryEquals(actual, expected);
+    }
+
+    private static void testJtsSerialization(String wkt)
+    {
+        Geometry jtsGeometry = createJtsGeometry(wkt);
+        OGCGeometry esriGeometry = OGCGeometry.fromText(wkt);
+
+        Slice jtsSerialized = JtsGeometrySerde.serialize(jtsGeometry);
+        Slice esriSerialized = GeometrySerde.serialize(esriGeometry);
+        assertEquals(jtsSerialized, esriSerialized);
+
+        Geometry jtsDeserialized = JtsGeometrySerde.deserialize(jtsSerialized);
+        assertGeometryEquals(jtsDeserialized, jtsGeometry);
+
+        OGCGeometry esriDeserialized = GeometrySerde.deserialize(esriSerialized);
+        assertGeometryEquals(esriDeserialized, esriGeometry);
+    }
+
+    private static Slice geometryFromText(String wkt)
+    {
+        return serialize(OGCGeometry.fromText(wkt));
+    }
+
+    private static Geometry createJtsGeometry(String wkt)
+    {
+        try {
+            return new WKTReader().read(wkt);
+        }
+        catch (ParseException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static void assertGeometryEquals(Geometry actual, Geometry expected)
+    {
+        if (expected.isEmpty()) {
```

Comment:
Since earlier commits enhanced serialization to avoid losing multi-types this test can be simplified to `assertEquals(actual.norm(), expected.norm());` in all cases.

Reply:
This actually caught a bug. Thanks.